File: backend_java/src/main/java/br/com/fiap/chronosdtn/controller/TransactionController.java
package br.com.fiap.chronosdtn.controller;

// Por que: Importa os DTOs de requisição e resposta para transações.
import br.com.fiap.chronosdtn.dto.TransactionRequest;
import br.com.fiap.chronosdtn.dto.TransactionResponse;
// Por que: Importa o serviço que processa a lógica de negócio das transações.
import br.com.fiap.chronosdtn.service.TransactionService;
// Por que: Importa o Jakarta Validation para validar anotações como @NotNull e @Positive.
import jakarta.validation.Valid;
// Por que: Importa classes HATEOAS para envelopar os dados em hypermedia.
import org.springframework.hateoas.CollectionModel;
import org.springframework.hateoas.EntityModel;
// Por que: Permite construir links dinâmicos apontando para métodos de controllers mapeados.
import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.*;
// Por que: Classes utilitárias de HTTP do Spring.
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

// Por que: Classes auxiliares de coleções.
import java.util.List;
import java.util.stream.Collectors;

// Por que: RestController sinaliza que a classe responde chamadas REST em formato JSON.
@RestController
// Por que: Mapeia a URL padrão para recursos de transações.
@RequestMapping("/api/transactions")
// Por que: Habilita CORS para conexões entre a API e aplicativos móveis.
@CrossOrigin(origins = "*", allowedHeaders = "*")
public class TransactionController {

    // Por que: Serviço contendo as regras de liquidação de saldos e enfileiramento.
    private final TransactionService transactionService;

    // Por que: Injeção de dependência via construtor.
    public TransactionController(TransactionService transactionService) {
        this.transactionService = transactionService;
    }

    // Por que: Retorna a listagem completa de transações financeiras cadastradas com suporte HATEOAS.
    @GetMapping
    public ResponseEntity<CollectionModel<EntityModel<TransactionResponse>>> getAllTransactions() {
        // Por que: Busca a lista de transações formatadas como DTO do serviço.
        List<EntityModel<TransactionResponse>> transactions = transactionService.getAllTransactions().stream()
                // Por que: Envolve cada DTO em um modelo EntityModel para anexar hypermedia.
                .map(response -> EntityModel.of(response,
                        // Por que: Cria o link HATEOAS de self-service da própria transação.
                        linkTo(methodOn(TransactionController.class).getTransactionById(response.transactionId())).withSelfRel(),
                        // Por que: Cria o link HATEOAS apontando para o bundle físico que transporta esta transação no espaço.
                        linkTo(methodOn(BundleController.class).getBundleById(response.sourceNodeId(), response.bundleId())).withRel("transport-bundle")
                ))
                // Por que: Coleta em lista.
                .collect(Collectors.toList());

        // Por que: Cria a coleção HATEOAS agregando a lista de transações e o link da própria busca.
        CollectionModel<EntityModel<TransactionResponse>> collectionModel = CollectionModel.of(transactions,
                linkTo(methodOn(TransactionController.class).getAllTransactions()).withSelfRel()
        );

        // Por que: Retorna status HTTP 200 OK com as entidades e links.
        return ResponseEntity.ok(collectionModel);
    }

    // Por que: Retorna o detalhamento individual de uma transação específica identificada por ID.
    @GetMapping("/{id}")
    public ResponseEntity<EntityModel<TransactionResponse>> getTransactionById(@PathVariable Long id) {
        // Por que: Busca os dados da transação.
        TransactionResponse response = transactionService.getTransactionById(id);

        // Por que: Cria o modelo HATEOAS com links para si mesmo e para o bundle de rede correspondente.
        EntityModel<TransactionResponse> entityModel = EntityModel.of(response,
                linkTo(methodOn(TransactionController.class).getTransactionById(id)).withSelfRel(),
                linkTo(methodOn(BundleController.class).getBundleById(response.sourceNodeId(), response.bundleId())).withRel("transport-bundle"),
                linkTo(methodOn(TransactionController.class).getAllTransactions()).withRel("all-transactions")
        );

        // Por que: Retorna HTTP 200 OK.
        return ResponseEntity.ok(entityModel);
    }

    // Por que: Mapeia requisições POST para iniciar a liquidação de créditos Terra-Lua.
    @PostMapping
    public ResponseEntity<EntityModel<TransactionResponse>> createTransaction(@Valid @RequestBody TransactionRequest request) {
        // Por que: Executa a criação física da transação no serviço, incluindo débitos/fila.
        TransactionResponse response = transactionService.createTransaction(request);

        // Por que: Envolve o retorno no EntityModel com os links dinâmicos para auditoria em tempo real.
        EntityModel<TransactionResponse> entityModel = EntityModel.of(response,
                linkTo(methodOn(TransactionController.class).getTransactionById(response.transactionId())).withSelfRel(),
                linkTo(methodOn(BundleController.class).getBundleById(response.sourceNodeId(), response.bundleId())).withRel("transport-bundle")
        );

        // Por que: Retorna HTTP 201 Created indicando que o recurso foi gravado no banco de dados.
        return ResponseEntity.status(HttpStatus.CREATED).body(entityModel);
    }
}

File: backend_net/ChronosDtnNetApi/Controllers/TransactionController.cs
// Por que: Importa namespaces do ASP.NET Core para endpoints de API, rotas, verbos HTTP e CORS.
using Microsoft.AspNetCore.Mvc;
// Por que: Importa namespace do ASP.NET Core para autenticação e controle de acesso com [Authorize].
using Microsoft.AspNetCore.Authorization;
// Por que: Importa namespaces do .NET para manipulação de listas e projeções com LINQ.
using System.Linq;
// Por que: Importa os DTOs de requisição e resposta da API.
using ChronosDtnNetApi.DTOs;
// Por que: Importa as interfaces de serviço de transações e bundles.
using ChronosDtnNetApi.Services;

// Por que: Organiza as controllers sob o namespace Controllers.
namespace ChronosDtnNetApi.Controllers;

// Por que: Restringe acesso a esta controller apenas para operadores autenticados via JWT.
[Authorize]
// Por que: Define a classe como ApiController de ASP.NET Core Web API.
[ApiController]
// Por que: Mapeia a rota padrão para "/api/transactions".
[Route("api/transactions")]
public class TransactionController : ControllerBase
{
    // Por que: Mantém a referência privada ao serviço de transações.
    private readonly ITransactionService _transactionService;

    // Por que: Construtor injetando o serviço correspondente.
    public TransactionController(ITransactionService transactionService)
    {
        // Por que: Inicializa a propriedade de serviço privada.
        _transactionService = transactionService;
    }

    // Por que: Endpoint GET /api/transactions que lista todas as transações com links HATEOAS.
    [HttpGet(Name = "GetAllTransactions")]
    public ActionResult GetAllTransactions()
    {
        // Por que: Recupera as transações do banco através do serviço.
        var transactions = _transactionService.GetAllTransactions();

        // Por que: Projeta cada transação no formato HAL JSON com links auto-descritivos.
        var hateoasTransactions = transactions.Select(t => new
        {
            // Por que: Copia as propriedades de dados da transação.
            t.TransactionId,
            t.BundleId,
            t.SourceNodeId,
            t.SourceAccountId,
            t.DestAccountId,
            t.Amount,
            t.SettlementStatus,
            t.TransactionTime,
            // Por que: Adiciona o nó de metadados de links da RFC de Hypermedia (HATEOAS).
            _links = new
            {
                // Por que: Link direto (self) para obter esta transação específica.
                self = new { href = Url.Link("GetTransactionById", new { id = t.TransactionId }) },
                // Por que: Link para o bundle físico associado que transporta a transação.
                transportBundle = new { href = Url.Link("GetBundleById", new { sourceNodeId = t.SourceNodeId, localSequenceId = t.BundleId }) }
            }
        }).ToList();

        // Por que: Cria o envelope de retorno com os links do recurso coleção.
        var response = new
        {
            _embedded = new
            {
                transactions = hateoasTransactions
            },
            _links = new
            {
                self = new { href = Url.Link("GetAllTransactions", null) }
            }
        };

        // Por que: Retorna HTTP 200 OK com o JSON contendo os dados e os links HATEOAS.
        return Ok(response);
    }

    // Por que: Endpoint GET /api/transactions/{id} que obtém detalhes de uma transação específica por ID com links HATEOAS.
    [HttpGet("{id}", Name = "GetTransactionById")]
    public ActionResult GetTransactionById(long id)
    {
        // Por que: Busca a transação no serviço.
        var t = _transactionService.GetTransactionById(id);

        // Por que: Constrói a resposta enriquecida com os links daquela transação individual.
        var response = new
        {
            t.TransactionId,
            t.BundleId,
            t.SourceNodeId,
            t.SourceAccountId,
            t.DestAccountId,
            t.Amount,
            t.SettlementStatus,
            t.TransactionTime,
            _links = new
            {
                self = new { href = Url.Link("GetTransactionById", new { id = t.TransactionId }) },
                transportBundle = new { href = Url.Link("GetBundleById", new { sourceNodeId = t.SourceNodeId, localSequenceId = t.BundleId }) },
                allTransactions = new { href = Url.Link("GetAllTransactions", null) }
            }
        };

        // Por que: Retorna HTTP 200 OK com a transação e seus hiperlinks.
        return Ok(response);
    }

    // Por que: Endpoint POST /api/transactions que recebe dados de transferência, debita saldo e enfileira pacotes.
    [HttpPost]
    public ActionResult CreateTransaction([FromBody] TransactionRequest request)
    {
        // Por que: Invoca o serviço para criar a transação.
        var t = _transactionService.CreateTransaction(request);

        // Por que: Constrói a resposta HAL correspondente para a nova transação criada.
        var response = new
        {
            t.TransactionId,
            t.BundleId,
            t.SourceNodeId,
            t.SourceAccountId,
            t.DestAccountId,
            t.Amount,
            t.SettlementStatus,
            t.TransactionTime,
            _links = new
            {
                self = new { href = Url.Link("GetTransactionById", new { id = t.TransactionId }) },
                transportBundle = new { href = Url.Link("GetBundleById", new { sourceNodeId = t.SourceNodeId, localSequenceId = t.BundleId }) }
            }
        };

        // Por que: Retorna HTTP 201 Created com a transação criada e a URL de acesso no cabeçalho Location.
        return CreatedAtRoute("GetTransactionById", new { id = t.TransactionId }, response);
    }
}
